TestStoreRangeMergeNonConsecutive loses Raft command on last range, times out

The following test appears to have failed:

[#2050](https://circleci.com/gh/cockroachdb/cockroach/2050):

```
I0502 15:30:39.768189     443 multiraft.go:694] node 100000001: group 2 raft ready
I0502 15:30:39.768254     443 multiraft.go:699] HardState updated: {Term:6 Vote:4294967297 Commit:16 XXX_unrecognized:[]}
I0502 15:30:39.769040     443 multiraft.go:702] New Entry[0]: 6/16 EntryNormal 0000000000000000799e447ae60f42c8: raft_id:2 origin_node_id:4294967297 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:52 > cmd_id:<wall_time:0 random:0 > key:"\00
I0502 15:30:39.769815     443 multiraft.go:705] Committed Entry[0]: 6/16 EntryNormal 0000000000000000799e447ae60f42c8: raft_id:2 origin_node_id:4294967297 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:52 > cmd_id:<wall_time:0 random:0 > key:"\00
I0502 15:30:39.771286     443 multiraft.go:873] dropping stale write to group 3
panic: test timed out after 5m0s

goroutine 3477 [running]:
testing.func·008()
	/usr/src/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/src/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x13466c8, 0x182a420, 0xa3, 0xa3, 0x1)
	/usr/src/go/src/testing/testing.go:556 +0xdc3
--
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:502 +0xd2a
github.com/cockroachdb/cockroach/util.func·005()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x80
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0x133
FAIL	github.com/cockroachdb/cockroach/storage	300.049s
=== RUN TestBatchBasics
I0502 15:31:31.356473     531 rocksdb.go:88] opening in-memory rocksdb instance
I0502 15:31:31.357485     531 rocksdb.go:113] closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
I0502 15:31:31.360548     531 rocksdb.go:88] opening in-memory rocksdb instance
I0502 15:31:31.360922     531 rocksdb.go:113] closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
I0502 15:31:31.364016     531 rocksdb.go:88] opening in-memory rocksdb instance
I0502 15:30:39.768189     443 multiraft.go:694] node 100000001: group 2 raft ready
I0502 15:30:39.768254     443 multiraft.go:699] HardState updated: {Term:6 Vote:4294967297 Commit:16 XXX_unrecognized:[]}
I0502 15:30:39.769040     443 multiraft.go:702] New Entry[0]: 6/16 EntryNormal 0000000000000000799e447ae60f42c8: raft_id:2 origin_node_id:4294967297 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:52 > cmd_id:<wall_time:0 random:0 > key:"\00
I0502 15:30:39.769815     443 multiraft.go:705] Committed Entry[0]: 6/16 EntryNormal 0000000000000000799e447ae60f42c8: raft_id:2 origin_node_id:4294967297 cmd:<internal_resolve_intent:<header:<timestamp:<wall_time:0 logical:52 > cmd_id:<wall_time:0 random:0 > key:"\00
I0502 15:30:39.771286     443 multiraft.go:873] dropping stale write to group 3
panic: test timed out after 5m0s

goroutine 3477 [running]:
testing.func·008()
	/usr/src/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/src/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x13466c8, 0x182a420, 0xa3, 0xa3, 0x1)
	/usr/src/go/src/testing/testing.go:556 +0xdc3
--
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:502 +0xd2a
github.com/cockroachdb/cockroach/util.func·005()
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:73 +0x80
created by github.com/cockroachdb/cockroach/util.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stopper.go:74 +0x133
FAIL	github.com/cockroachdb/cockroach/storage	300.049s
=== RUN TestBatchBasics
I0502 15:31:31.356473     531 rocksdb.go:88] opening in-memory rocksdb instance
I0502 15:31:31.357485     531 rocksdb.go:113] closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN TestBatchGet
I0502 15:31:31.360548     531 rocksdb.go:88] opening in-memory rocksdb instance
I0502 15:31:31.360922     531 rocksdb.go:113] closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN TestBatchMerge
I0502 15:31:31.364016     531 rocksdb.go:88] opening in-memory rocksdb instance

```
Please assign, take a look and update the issue accordingly.

--- a/multiraft/multiraft.go
+++ b/multiraft/multiraft.go
@@ -654,10 +654,21 @@ func (s *state) createGroup(groupID uint64) error {
 
 func (s *state) removeGroup(op *removeGroupOp) {
 	// Group creation is lazy and idempotent; so is removal.
-	if _, ok := s.groups[op.groupID]; !ok {
+	log.V(3).Infof("removing group %d", op.groupID)
+	g, ok := s.groups[op.groupID]
+	if !ok {
 		op.ch <- nil
 		return
 	}
+
+	// Cancel commands which are still in transit.
+	for cmdID, prop := range g.pending {
+		if prop.ch != nil {
+			prop.ch <- util.Errorf("group deleted")
+		}
+		delete(g.pending, cmdID)
+	}
+
 	if err := s.multiNode.RemoveGroup(op.groupID); err != nil {
 		op.ch <- err
 		return
@@ -684,6 +695,7 @@ func (s *state) propose(p *proposal) {
 		}
 		return
 	}
+	log.V(3).Infof("group %d: new proposal %x", p.groupID, p.commandID)
 	g.pending[p.commandID] = p
 	p.fn()
 }

--- a/storage/client_merge_test.go
+++ b/storage/client_merge_test.go
@@ -198,10 +198,8 @@ func TestStoreRangeMergeLastRange(t *testing.T) {
 
 // TestStoreRangeMergeNonConsecutive attempts to merge two ranges
 // that are not on same store.
-// TODO(tschottdorf): debug the test failures on CircleCI, see #893.
 func TestStoreRangeMergeNonConsecutive(t *testing.T) {
 	defer leaktest.AfterTest(t)
-	t.Skip("TODO(tschottdorf): debug and re-enable, see #893")
 	store, stopper := createTestStore(t)
 	defer stopper.Stop()
 

--- a/storage/store.go
+++ b/storage/store.go
@@ -539,8 +539,8 @@ func (s *Store) startGossip() error {
 func (s *Store) maybeGossipFirstRange() error {
 	rng := s.LookupRange(engine.KeyMin, engine.KeyMin.Next())
 	if rng != nil {
-		log.V(1).Infof("store %d has first range, maybe gossiping",
-			s.StoreID())
+		log.Infof("gossiping first range on store %d, range %d",
+			s.StoreID(), rng.Desc().RaftID)
 		return rng.maybeGossipFirstRange()
 	}
 	return nil
@@ -558,10 +558,11 @@ func (s *Store) maybeGossipConfigs() error {
 	for _, cd := range configDescriptors {
 		rng := s.LookupRange(cd.keyPrefix, cd.keyPrefix.Next())
 		if rng == nil {
-			log.Warningf("no range")
 			// This store has no range with this configuration.
 			continue
 		}
+		log.Infof("gossiping configuration map on store %d, range %d: %s",
+			s.StoreID(), rng.Desc().RaftID, cd.gossipKey)
 		// Wake up the replica. If it acquires a fresh lease, it will
 		// gossip. If an unexpected error occurs (i.e. nobody else seems to
 		// have an active lease but we still failed to obtain it), return
@@ -1247,6 +1248,10 @@ func (s *Store) processRaft() {
 					err := gogoproto.Unmarshal(e.Command, &cmd)
 					if err != nil {
 						log.Fatal(err)
+					}
+					if log.V(6) {
+						log.Infof("store %d: new committed %s command at index %d",
+							cmd.Cmd.GetValue().(proto.Request).Method(), e.Index)
 					}
 
 				case *multiraft.EventMembershipChangeCommitted:

